# Question 1 - Supply Chain Optimization under  Demand Uncertainty.

In [ ]:
import numpy as np
import cvxpy as cp

In [ ]:
# Decision variable
#decision variable x represents the supply
x = cp.Variable()

In [ ]:
# Demand scenarios and probabilities
demand_scenarios = np.array([255, 302, 270, 317, 285, 332, 300, 347, 315, 362, 330, 262, 309, 277])
demand_probabilities = np.array([0.03, 0.15, 0.04, 0.1, 0.05, 0.05, 0.09, 0.01, 0.09, 0.03, 0.2, 0.07, 0.05, 0.04])

In [ ]:
# Cost function (minimize expected cost)
cost = cp.sum(cp.multiply(demand_probabilities, cp.pos(x - demand_scenarios)))
objective = cp.Minimize(cost)

In [ ]:
# Constraints
constraints = [x >= 120, x <= 330]

In [ ]:
# Solve problem
problem = cp.Problem(objective, constraints)
problem.solve()


0.0

In [ ]:
print(f"Optimal Supply Level: {x.value}")

Optimal Supply Level: 249.3443890640451


**Optimal Supply Level: 249.34**

The optimal supply level of 249.34 suggests that the best strategy the company can use to minimize the expected cost under uncertain demand scenarios (ranging from 255 to 362 units) is to set the supply level at approximately 249.34 units.

**Underproduction Costs**: The optimal supply level of 249.34 is less than most of the demand scenarios in the supply chain. Therefore, there will be a shortfall in supply for most scenarios when the supply level is set to 249.34. This suggests that it is more cost-effective to underproduce and accept penalties for unmet demand rather than overproduce and incur holding costs.

**Expected Cost Minimization**: The model finds the balance where the expected penalty cost (due to demand exceeding supply) is optimized (minimized in this case). The decision variable supply level (x) is positioned optimally given the probabilities of different levels of demand.

**Practical Implication**: In a real-world supply chain setting, this solution implies that producing or stocking electronics around 249.34 units is the best choice, given the stochastic nature of demand and the assumed cost function.

# Question 2 - Healthcare Resource Allocation with Uncertain Demand.

In [ ]:
import numpy as np
import cvxpy as cp

In [ ]:
# Decision variable
beds = cp.Variable()

In [ ]:
# Patient arrival scenarios and probabilities
arrival_scenarios = np.array([25, 20, 30, 50, 27, 39, 42, 29, 35, 42])
arrival_probabilities = np.array([0.1, 0.1, 0.08, 0.15, 0.09, 0.05, 0.1, 0.1, 0.13, 0.1])

In [ ]:
# Cost function (minimize expected cost)
cost = cp.sum(cp.multiply(arrival_probabilities, cp.pos(beds - arrival_scenarios)))
objective = cp.Minimize(cost)

In [ ]:
# Constraints
constraints = [beds >= 17, beds <= 55]

In [ ]:
# Solve problem
problem = cp.Problem(objective, constraints)
problem.solve()

0.0

In [ ]:
print(f"Optimal number of ICU beds: {beds.value}")

Optimal number of ICU beds: 19.628282074057484


**Optimal Number of ICU Beds: 19.63**

The optimal number of ICU beds of 19.63 suggests that the best strategy to minimize the expected cost under uncertain patient arrival scenarios (ranging from 20 to 50 patients) is to allocate approximately 19.63 beds.

**Underallocation Costs**: The optimal number of beds is 19.63, which is less than most of the patient arrival scenarios shuch as 20, 25, 30. This, there will be a shortfall in bed allocation for most scenarios. This implies that it is more cost-effective to underallocate beds and accept penalties for unmet demand rather than overallocate and incur unnecessary costs. This decision will accomodate most of the patient demand scenarios while minimizing excess bed capacity.

**Expected Cost Minimization**: The model finds the balance where the expected penalty cost (due to patient arrivals exceeding bed capacity) is minimized. The decision variable beds is positioned optimally given the probabilities of different patient arrival levels.

**Practical Implication**: In a real-world healthcare setting, this solution suggests that allocating around 19.63 ICU beds is the best choice, given the stochastic/random nature of patient arrivals and the assumed cost function.